tavern/internal/redirectors/dns/dns.go

```diff
@@ -0,0 +1,1094 @@
+			end := i + maxResponseChunkSize
+			if end > len(encodedResponse) {
+				end = len(encodedResponse)
+			}
+			conv.ResponseChunks = append(conv.ResponseChunks, encodedResponse[i:end])
+		}
+
+		// Return chunked response indicator: "r:[num_chunks]:[crc]"
+		response := fmt.Sprintf("%s%s:%s", respChunked, encodeSeq(len(conv.ResponseChunks)), EncodeBase36CRC(int(conv.ResponseCRC)))
+		slog.Debug("returning chunked response indicator", "conv_id", convID, "chunks", len(conv.ResponseChunks), "crc", conv.ResponseCRC)
+		return []byte(response), nil
+	}
+
+	return []byte(responseWithPrefix), nil
+}
+
+// HandleFetchPacket serves a response chunk to the client
+func (r *Redirector) HandleFetchPacket(convID string, chunkSeq int) ([]byte, error) {
+	convVal, ok := r.conversations.Load(convID)
+	if !ok {
+		return nil, fmt.Errorf("unknown conversation: %s", convID)
+	}
+
+	conv := convVal.(*Conversation)
+	conv.mu.Lock()
+	defer conv.mu.Unlock()
+
+	conv.LastActivity = time.Now()
+
+	// Check if this is the final fetch (cleanup request)
+	if chunkSeq >= len(conv.ResponseChunks) {
+		// Client is done fetching - clean up conversation
+		r.conversations.Delete(convID)
+		slog.Debug("conversation completed and cleaned up", "conv_id", convID)
+		return []byte(respOK), nil
+	}
+
+	// Return the requested chunk
+	if chunkSeq < 0 || chunkSeq >= len(conv.ResponseChunks) {
+		return nil, fmt.Errorf("invalid chunk sequence: %d (total: %d)", chunkSeq, len(conv.ResponseChunks))
+	}
+
+	chunk := conv.ResponseChunks[chunkSeq]
+	slog.Debug("serving response chunk", "conv_id", convID, "seq", chunkSeq, "size", len(chunk), "is_binary", conv.IsBinaryChunking)
+
+	// For binary chunking (A/AAAA), return raw bytes
+	// For text chunking (TXT), return "ok:" prefix + base32 data
+	if conv.IsBinaryChunking {
+		return []byte(chunk), nil
+	}
+	return []byte(respOK + chunk), nil
+}
+
+// reassembleChunks combines chunks in order
+func (r *Redirector) reassembleChunks(chunks map[int][]byte, totalChunks int) []byte {
+	var result []byte
+	for i := 0; i < totalChunks; i++ {
+		if chunk, ok := chunks[i]; ok {
+			slog.Debug("reassembling chunk", "seq", i, "chunk_len", len(chunk), "total_so_far", len(result))
+			result = append(result, chunk...)
+		} else {
+			// This should never happen since we check for missing chunks first
+			slog.Error("CRITICAL: Missing chunk during reassembly", "seq", i, "total_chunks", totalChunks, "chunks_present", len(chunks))
+		}
+	}
+	slog.Debug("reassembly complete", "final_len", len(result), "total_chunks", totalChunks)
+	return result
+}
+
+// forwardToUpstream sends request to gRPC server and returns response
+func (r *Redirector) forwardToUpstream(ctx context.Context, upstream *grpc.ClientConn, methodPath string, requestData []byte) ([]byte, error) {
+	// Create gRPC stream
+	md := metadata.New(map[string]string{})
+	ctx = metadata.NewOutgoingContext(ctx, md)
+
+	stream, err := upstream.NewStream(ctx, &grpc.StreamDesc{
```

Comment:
Try looking into the grpc stream size variable options.
I'm pretty sure you can set the message size to something that would fit into a DNS request / response and then you wouldn't need to do chunkning by hand.